src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala

```diff
@@ -60,34 +62,38 @@ object InsertMemoryBarriers {
                              metadata: Metadata) extends VisitAndRebuild.Visitor {
     override def phrase[T <: PhraseType](p: Phrase[T]): Result[Phrase[T]] = {
       p match {
-        case For(n, Lambda(x, body), unroll) =>
-          Stop(For(n, Lambda(x, visitLoopBody(body, allocs, metadata)), unroll))
-        case ForNat(n, DepLambda(x, body), unroll) =>
-          Stop(ForNat(n, DepLambda(x, visitLoopBody(body, allocs, metadata)), unroll))
-        case pf@OpenCLParFor(n, dt, out, Lambda(x, Lambda(o, body)), init, step, unroll) =>
-          pf match {
-            case ParForLocal(dim) =>
-              val outer_wg_writes = mutable.Map[Identifier[_ <: PhraseType], AddressSpace]()
-              collectWrites(out, allocs, outer_wg_writes)
-              val b2 = visitLoopBody(body, allocs, metadata, outer_wg_writes)
-              Stop(ParForLocal(dim)(n, dt, out, Lambda(x, Lambda(o, b2)), init, step, unroll))
-            case ParForWorkGroup(dim) =>
-              Stop(ParForWorkGroup(dim)(n, dt, out,
-                Lambda(x, Lambda(o, visitLoopBody(body, allocs, metadata))), init, step, unroll))
-            case ParForGlobal(dim) =>
-              Stop(ParForGlobal(dim)(n, dt, out,
-                Lambda(x, Lambda(o, visitLoopBody(body, allocs, metadata))), init, step, unroll))
-          }
-        case pfn: OpenCLParForNat => ???
-        case OpenCLNew(addr, _, Lambda(x, _)) if addr != AddressSpace.Private =>
+        case f@For(unroll) =>
+          val (x, body) = f.unwrapBody
+          Stop(For(unroll)(f.n, Lambda(x, visitLoopBody(body, allocs, metadata))))
+        case f@ForNat(unroll) =>
+          val (x, body) = f.unwrapBody
+          Stop(ForNat(unroll)(f.n,
+            DepLambda[NatKind, CommType](x, visitLoopBody(body, allocs, metadata))))
+        case pf@ParFor(Local, dim, unroll) =>
+          val (x, o, body) = pf.unwrapBody
+          val outer_wg_writes = mutable.Map[Identifier[_ <: PhraseType], AddressSpace]()
+          collectWrites(pf.out, allocs, outer_wg_writes)
+          Stop(ParFor(Local, dim, unroll)(pf.n, pf.dt, pf.out,
+            Lambda(x, Lambda(o,
+              visitLoopBody(body, allocs, metadata, outer_wg_writes))), pf.init, pf.step))
+        case pf@ParFor(level, dim, unroll) =>
+          val (x, o, body) = pf.unwrapBody
+          Stop(ParFor(level, dim, unroll)(pf.n, pf.dt, pf.out,
+            Lambda(x, Lambda(o, visitLoopBody(body, allocs, metadata))), pf.init, pf.step))
+        case pf@ParForNat(level, dim, unroll) =>
+          val (x, o, body) = pf.unwrapBody
+          Stop(ParForNat(level, dim, unroll)(pf.n, pf.ft, pf.out,
+            DepLambda[NatKind, AccType ->: CommType](x,
+              Lambda(o, visitLoopBody(body, allocs, metadata))), pf.init, pf.step))
```

Comment:
looks like the previous `ParForNat(Local` case is missing.

Reply:
This wasn't implemented before (just a `???`) but this case should be analogous to the `ParFor(Local` case. Which indeed seems missing. I'll add it.